frontend/src/q.ts

```diff
@@ -1,13 +1,24 @@
-import { queryOptions } from "@tanstack/react-query";
-import lunr from "lunr";
-import { api } from "./query";
+import { queryOptions, skipToken } from "@tanstack/react-query";
+import { api } from "@/query";
+import { definitions } from "./api";
 
-export const getSearchIndexQuery = () =>
+export const getSearchQuery = (query: string) =>
   queryOptions({
-    queryKey: ["search-index"],
-    queryFn: async () => {
-      const data = await api(`search.json`).json();
+    queryKey: ["search", query],
+    queryFn:
+      query.length > 0
+        ? async ({ signal }) => {
+            // This is the lazy man's debounce, by waiting 100ms before making the request
+            // and utilizing the signal to cancel the request if the query changes
+            // we get a debounce effect
+            await (() => new Promise((resolve) => setTimeout(resolve, 100)))();
+            if (signal.aborted) {
+              return;
+            }
 
-      return lunr.Index.load(data);
-    },
+            return await api(`search?q=${encodeURIComponent(query)}`, {
+              signal,
+            }).json<definitions["SearchResultItem"][]>();
+          }
+        : skipToken,
```

Comment:
Shouldn't this approach send a request for every character you type, even if you do it faster than the debounce (100ms)?

It should probably be better to debounce the query key - https://stackoverflow.com/questions/76030996/how-to-use-debounce-with-usequery-in-react-query

Reply:
If we want to debounce it I'd suggest moving this out of the query and replacing the `useDeferredValue` (in the Search component) with a simple hook that would do the debouncing. We could either use something from npm or do something simple like:
```ts
function useDebouncedValue(value: string, delay: number) {
  const [debouncedValue, setDebouncedValue] = useState(value);

  useEffect(() => {
    const handler = setTimeout(() => {
      setDebouncedValue(value);
    }, delay);

    return () => {
      clearTimeout(handler);
    };
  }, [value, delay]);

  return debouncedValue;
}
```

Reply:
I was under the assumption that the tanstack query signal that gets sent down handles the cancellation. By me checking `if (signal.aborted)` im checking if tanstack aborted the query so i can move on. 

But I could be wrong in these assumptions.

Reply:
In theory `fetch` shouldn't even start if a signal is already aborted.

Reply:
Thank you, wasn't aware of this